### PR TITLE
Add support for weighted StatsBase methods

### LINKED ISSUE #26: Wrappers for StatsBase
I realize this is introducing an extra dependency, but it'd be nice if you could do things like:

```
μ, σ = mean_and_std(A, wv, :channel; corrected=true)
```

### PR DESCRIPTION
**Pull request by rofinn**
This PR doesn't wrap everything, but it wraps most of the main weighted functions that we use at Invenia. Closes #26

### CODE REVIEW

**Comment by mcabbott** [comment]
This looks good, thanks. 

I have one design question. In some sense a natural place for `AbstractWeights` to live is as one of the key-vectors, attached to a dimension. Is this something which ought to be made easy? I'm not sure how, probably it would it be too confusing for `mean(A)` to behave completely differently for `axiskeys(A,2) isa AbstractWeights` compared to a vector, and you may also want `mean(A, dims=2)` to ignore the keys if they have a different meaning.

**Comment by rofinn** [comment]
Yeah, I feel like having the keys be statistical weights would be a bit confusing... at least for our use cases. We typically want to do things like compute the exponentially weighted mean along a `:time` dimension which would already have ordered `DateTime` keys. I suppose if you could embed the weights alongside the existing keys then that would be kinda slick, but I'm not sure if that's worth the extra complications. I think a more useful addition might be to support wrapped weights. For example:

```
mean(A::KeyedArray, wv::KeyedVector{T, <:AbstractWeights}; kwargs...)
```

That could still be useful for seeing how weights map to each key along that dimension.

**Comment by rofinn** [comment]
Alright, I've rebased and added support for KeyedArray weight vectors. I've also opted to wrap the covariance estimation methods related to a NameDims.jl issue. https://github.com/invenia/NamedDims.jl/issues/120

**Comment by rofinn** [comment]
Queued job seems to have completed, but github wasn't notified.

**Comment by mcabbott** [comment]
I dropped the ball here too, sorry.

Re using key-vector as weights, maybe I start to agree that this isn't such a great idea. 

Is CovarianceEstimation a fairly lightweight package? I guess if https://github.com/invenia/NamedDims.jl/pull/122 loads it, then there's no cost to loading it here too. While if NamedDims uses Requires.jl, then possibly this should follow that?

**Comment by rofinn** [comment]
Yeah, CovarianceEstimation.jl would only add StatsBase.jl, which we're considering adding here anyways. My issue with Requires.jl is that the load time performance hasn't been great (might be better on julia >= 1.5 though). It also doesn't let you specify dependency requirements which is particularly problematic for packages like StatsBase and CovarianceEstimation which are still pre-1.0.

- https://github.com/JuliaPackaging/Requires.jl/issues/39
- https://github.com/JuliaPackaging/Requires.jl/issues/82

**Comment by mcabbott** [comment]
OK, I think Requires has got faster, but haven't timed things exhaustively. It is loaded by NamedDims already, but I also don't know whether that influences things. I am a little surprised how many things StatsBase depends on, despite its name: https://juliahub.com/ui/Packages/StatsBase/EZjIG/0.33.1?t=1

**Comment by rofinn** [comment]
> I am a little surprised how many things StatsBase depends on, despite its name

Yeah, that's fair. I think the plan is for a lot of the core functionality to move into the Statistics stdlib? That being said, it's only 4 non-stdlibs and those package are minimal enough to not have further indirect dependencies. I guess my comparison is packages like Flux.jl https://juliahub.com/ui/Packages/Flux/QdkVy/0.11.1?t=1

**Comment by rofinn** [comment]
Sorry, I got confused. Is there anything in particular you wanted me to fix with this PR (e.g., use Requires.jl)?

**Comment by mcabbott** [comment]
Have just been busy, sorry. Will look properly soon.

File: src/statsbase.jl [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        rem_name = dimnames(A, rem_dim)
+        rem_key = axiskeys(A, rem_dim)
+        return KeyedArray(
+            NamedDimsArray(data, (rem_name, rem_name)),
+            (copy(rem_key), copy(rem_key))
+        )
+    else
+        data = cov(parent(A), wv, dims; corrected=corrected, kwargs...)
+        # Use same remaining axis for both dimensions of data
+        rem_dim = first(setdiff(ndims(A), dims))
```
Comment (mcabbott) [inline src/statsbase.jl:53]: Can we avoid this?

File: src/statsbase.jl [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        rem_dim = first(setdiff((1, 2), dims))
+        new_keys = Tuple(copy(axiskeys(A, rem_dim)) for i in 1:2)
+        return KeyedArray(data, new_keys)
+    end
+end
+
+function Statistics.cor(A::KeyedArray, wv::KeyedVector; kwargs...)
+    return cor(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+end
+
```
Comment (mcabbott) [inline src/statsbase.jl:89]:
```suggestion

```

File: test/_packages.jl [hunk 3]
```diff
@@ -134,3 +134,107 @@ end
...
 end
+
+@testset "statsbase" begin
+    using CovarianceEstimation, StatsBase
+
+    A = rand(4, 3)
+    A_ka = KeyedArray(A, (0:3, [:a, :b, :c]))
+    A_kanda = KeyedArray(A; time = 0:3, id = [:a, :b, :c])
+    wv = aweights(rand(4))
+    kwv = KeyedArray(wv; time = 0:3)
```
Comment (mcabbott) [inline test/_packages.jl:145]:
```suggestion
```

File: src/statsbase.jl [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        rem_dim = first(setdiff(ndims(A), dims))
+        new_keys = Tuple(copy(axiskeys(A, rem_dim)) for i in 1:2)
+        return KeyedArray(data, new_keys)
+    end
+end
+
+function Statistics.cov(A::KeyedArray, wv::KeyedVector; kwargs...)
+    return cov(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+end
+
```
Comment (mcabbott) [inline src/statsbase.jl:62]:
```suggestion
```

File: src/statsbase.jl [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        new_keys = ntuple(d -> d in numerical_dims ? Base.OneTo(1) : axiskeys(A,d), ndims(A))
+        return KeyedArray(data, map(copy, new_keys))#, copy(A.meta))
+    end
+
+    @eval function Statistics.$fun(A::KeyedArray, wv::KeyedVector; kwargs...)
+        return $fun(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+    end
+end
+
+function Statistics.cov(A::KeyedMatrix, wv::AbstractWeights; dims=1, corrected=true, kwargs...)
```
Comment (mcabbott) [inline src/statsbase.jl:36]: Can this be combined with `Statistics.cor` in an `@eval` loop? It looks like the only difference is the keyword `corrected=true`, but won't this be passed along as part of `kwargs...` anyway?

File: src/statsbase.jl [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+    # A little awkward, but the weighted `cov` method from statsbase only works
+    # on dense matrices, so we need to unwrap this twice and completely rebuild the
+    # array
+    if AxisKeys.hasnames(A)
+        numerical_dim = NamedDims.dim(dimnames(A), dims)
+        data = cov(parent(parent(A)), wv, numerical_dim; corrected=corrected, kwargs...)
+        rem_dim = first(setdiff((1, 2), numerical_dim))
```
Comment (mcabbott) [inline src/statsbase.jl:43]: I think these only accept `dims::Int`. In which case this can just be `3 - numerical_dim`.

File: src/statsbase.jl [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            NamedDimsArray(data, (rem_name, rem_name)),
+            (copy(rem_key), copy(rem_key))
+        )
+    else
+        data = cor(parent(A), wv, dims; kwargs...)
+        # Use same remaining axis for both dimensions of data
+        rem_dim = first(setdiff((1, 2), dims))
+        new_keys = Tuple(copy(axiskeys(A, rem_dim)) for i in 1:2)
+        return KeyedArray(data, new_keys)
+    end
```
Comment (mcabbott) [inline src/statsbase.jl:83]: There has got to be a more compact pattern for these things. Maybe it's like this, will this work? 
```suggestion
    d = NamedDims.dim(dimnames(A), dims)
    data = scattermat(unname(keyless(A)), wv...; dims=d, kwargs...)
    L1 = dimnames(A, 3-d)
    data2 = hasnames(A) ? NamedDimsArray(data, (L1, L1)) : data
    K1 = axiskeys(A, 3-d)
    KeyedArray(data2, (copy(K1), copy(K1)))
```
Perhaps there should be a unified function for doing `unname(keyless`, but probably still pretty cheap.

File: src/statsbase.jl [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+end
+
+function StatsBase.scattermat(A::KeyedArray, wv::KeyedVector; kwargs...)
+    return scattermat(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+end
+
```
Comment (mcabbott) [inline src/statsbase.jl:117]:
```suggestion
```

File: src/statsbase.jl [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+function StatsBase.mean_and_std(A::KeyedMatrix, wv::Vararg{<:AbstractWeights}; dims=:, corrected::Bool=true, kwargs...)
+    return (
+        mean(A, wv...; dims=dims, kwargs...),
+        std(A, wv...; dims=dims, corrected=corrected, kwargs...)
+    )
+end
+function StatsBase.mean_and_std(A::KeyedArray, wv::KeyedVector; kwargs...)
+    return mean_and_std(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+end
```
Comment (mcabbott) [inline src/statsbase.jl:126]:
```suggestion
```

File: src/statsbase.jl [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    return (
+        mean(A, wv...; dims=dims, kwargs...),
+        var(A, wv...; dims=dims, corrected=corrected, kwargs...)
+    )
+end
+function StatsBase.mean_and_var(A::KeyedArray, wv::KeyedVector; kwargs...)
+    return mean_and_var(A, hasnames(wv) ? parent(parent(wv)) : parent(wv); kwargs...)
+end
+
+function StatsBase.mean_and_cov(A::KeyedMatrix, wv::Vararg{<:AbstractWeights}; dims=:, corrected::Bool=true, kwargs...)
```
Comment (mcabbott) [inline src/statsbase.jl:138]: Can these `mean_and_x` funtions also be done in an `@eval` loop?

**Review by mcabbott — CHANGES_REQUESTED** [review]
Sorry this took forever. 

I think this is fine but we should compact some things, by making some more loops, by shortening some bodies, and by deleting some methods.

File: src/statsbase.jl (hunk 7 above)
Comment (mcabbott) [reply src/statsbase.jl:83]: I made one here: https://github.com/mcabbott/AxisKeys.jl/commit/82d7037de5e99c5e2d48594de76dd3b0780c83c2

File: src/statsbase.jl (hunk 5 above)
Comment (rofinn) [reply src/statsbase.jl:36]: Yeah, it's just that the default behaviour in StatsBase will throw a deprecation warning stating that `corrected=true` will be the default in a future release. If we're okay with the deprecation in our tests then I can merge them and the warning will go away in a future release.

File: src/statsbase.jl (hunk 10 above)
Comment (rofinn) [reply src/statsbase.jl:138]: Yeah, I think if I just build up the full func name in the loop then we should be good.

File: src/statsbase.jl (hunk 5 above)
Comment (mcabbott) [reply src/statsbase.jl:36]: Oh I see. I guess I don't mind the noise, and eventually it will stop as you say.

Comment (rofinn) [reply src/statsbase.jl]: N/m, just changed the tests to explicitly pass `corrected=true` to avoid the noise.

**Comment by rofinn** [comment]
Alright, looks like that managed to remove about 100 lines of largely duplicated code.

File: src/statsbase.jl [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    :DiagonalUnitVariance,
+    :DiagonalCommonVariance,
+    :DiagonalUnequalVariance,
+    :CommonCovariance,
+    :PerfectPositiveCorrelation,
+    :ConstantCorrelation,
+    :AnalyticalNonlinearShrinkage,
+]
+for estimator in estimators
+    @eval function Statistics.cov(ce::$estimator, A::KeyedMatrix, wv::Vararg{<:AbstractWeights}; dims=:, kwargs...)
```
Comment (mcabbott) [inline src/statsbase.jl:75]: Should this have `dims=1`?

File: src/functions.jl [hunk 12]
```diff
@@ -45,23 +45,37 @@ end
...
 end
 
+for fun in [:cov, :cor] # Returned the axes work are different for cov and cor
+    @eval function Statistics.$fun(A::KeyedMatrix; dims=1, kwargs...)
+        numerical_dim = hasnames(A) ? NamedDims.dim(dimnames(A), dims) : dims
+        data = $fun(parent(A); dims=numerical_dim, kwargs...)
+        # Use same remaining axis for both dimensions of data
+        rem_dim = numerical_dim == 1 ? 2 : 1
+        new_keys = Tuple(copy(axiskeys(A, rem_dim)) for i in 1:2)
+        return KeyedArray(data, new_keys)
```
Comment (mcabbott) [inline src/functions.jl:75]:
```suggestion
        rem_key = axiskeys(A, 3-numerical_dim)
        KeyedArray(data, (copy(rem_key), copy(rem_key)))
```

File: src/statsbase.jl (hunk 11 above)
Comment (rofinn) [reply src/statsbase.jl:75]: Yeah, I guess `dims=:` only makes sense for function like `mean`.

Comment (rofinn) [reply src/statsbase.jl:75]:
```suggestion
    @eval function Statistics.cov(ce::$estimator, A::KeyedMatrix, wv::Vararg{<:AbstractWeights}; dims=1, kwargs...)
```

**Pull request merged**